cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/AppFabricProcessorServiceMain.java

```diff
@@ -0,0 +1,184 @@
+import io.cdap.cdap.data2.datafabric.dataset.service.executor.DatasetOpExecutorService;
+import io.cdap.cdap.data2.metadata.writer.DefaultMetadataServiceClient;
+import io.cdap.cdap.data2.metadata.writer.MessagingMetadataPublisher;
+import io.cdap.cdap.data2.metadata.writer.MetadataPublisher;
+import io.cdap.cdap.data2.metadata.writer.MetadataServiceClient;
+import io.cdap.cdap.internal.app.namespace.LocalStorageProviderNamespaceAdmin;
+import io.cdap.cdap.internal.app.namespace.StorageProviderNamespaceAdmin;
+import io.cdap.cdap.internal.app.services.AppFabricProcessorService;
+import io.cdap.cdap.internal.app.worker.TaskWorkerServiceLauncher;
+import io.cdap.cdap.internal.app.worker.system.SystemWorkerServiceLauncher;
+import io.cdap.cdap.internal.events.EventPublishManager;
+import io.cdap.cdap.master.spi.environment.MasterEnvironment;
+import io.cdap.cdap.master.spi.environment.MasterEnvironmentContext;
+import io.cdap.cdap.messaging.guice.MessagingServiceModule;
+import io.cdap.cdap.metrics.guice.MetricsStoreModule;
+import io.cdap.cdap.operations.OperationalStatsService;
+import io.cdap.cdap.operations.guice.OperationalStatsModule;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.security.authorization.AccessControllerInstantiator;
+import io.cdap.cdap.security.authorization.AuthorizationEnforcementModule;
+import io.cdap.cdap.security.guice.SecureStoreServerModule;
+import io.cdap.cdap.security.store.SecureStoreService;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import org.apache.twill.api.TwillRunner;
+import org.apache.twill.api.TwillRunnerService;
+import org.apache.twill.zookeeper.ZKClientService;
+
+/**
+ * The main class to run appfabric processor and other supporting services.
+ */
+public class AppFabricProcessorServiceMain extends AbstractServiceMain<EnvironmentOptions> {
+
+  /**
+   * Main entry point.
+   */
+  public static void main(String[] args) throws Exception {
+    main(AppFabricProcessorServiceMain.class, args);
+  }
+
+  @Override
+  protected List<Module> getServiceModules(MasterEnvironment masterEnv,
+      EnvironmentOptions options, CConfiguration cConf) {
+    return Arrays.asList(
+        // Always use local table implementations, which use LevelDB.
+        // In K8s, there won't be HBase and the cdap-site should be set to use SQL store for StructuredTable.
+        new DataSetServiceModules().getStandaloneModules(),
+        // The Dataset set modules are only needed to satisfy dependency injection
+        new DataSetsModules().getStandaloneModules(),
+        new MetricsStoreModule(),
+        new MessagingServiceModule(cConf),
+        new AuditModule(),
+        new AuthorizationModule(),
+        new AuthorizationEnforcementModule().getMasterModule(),
+        Modules.override(new AppFabricServiceRuntimeModule(cConf).getDistributedModules())
```

Comment:
Are you using same runtime module for Service and Processor? How does it decide what to bring up where?

Reply:
Yes, same module is used.

`AppFabricServer` and `AppFabricProcessorService` decide what to bring up for their service.

Reply:
I feel it's a bit too easy to mess up. We define both things and then if anyone refers to a thing with a different name, we get all the handlers up... It's probably not the biggest deal as Processor handlers won't be able to come up without necessary services, but it also introduces a hidden assumption.
May I ask you to add some explicit thing, e.g. add a parameter to AppFabricServiceRuntimeModule constructor (PROCESSOR_HANDLERS/SERVER_HANDLERS/BOTH_HANDLERS) and bind only the handlers this module is constructed for? This would make it explicit and ensures it's hard to mess it up.

Reply:
SG